todo_app/app.py

```diff
@@ -3,69 +3,168 @@
+            
+            if "login" not in r.json():
+                # HELP - is there a better way to handle this? how to return 401 - Unauthorized?
+                raise Exception("Authentication Failed at Step3. See application logs.")
+            user_id = r.json()["login"]
+                        
+            app.logger.info(f"Logging in User: {user_id}")
+            user = User(user_id)
+            logged_in = login_user(user)          
+
             return redirect('/')
+
         except Exception as e:
+            # HELP - is there a better way to handle this? how to return 401 - Unauthorized?
             app.logger.error("Error: %s", e)
             raise e
 
+    @app.route('/tasks', methods=['POST'])
+    @login_required
+    def add_new_task():
+        if (current_user.get_role() == UserRole.writer):
+            try:
+                app.logger.info("add_new_task()")
+                task_name = request.form["task_name"]
+                app.logger.info(f"add_new_task() => {task_name}")
+                repository.add_task(task_name, "ToDo")
+                return redirect('/')
+            except Exception as e:
+                app.logger.error("Error: %s", e)
+                raise e
+        else:
+            # HELP - is there a better way to handle this? how to return 401 - Unauthorized?
+            raise Exception("Unauthorized. To Add a New Task, the User must have the 'writer' role.")
+
     @app.route('/change_status/<id>', methods=['POST']) 
+    @login_required
     def change_status(id): 
-        try:
-            status = request.args["status"]
-            app.logger.info(f"mark_as_done: id = {id}, status = {status}")
-            repository.update_task_status(id, status)
-            return redirect('/')
-        except Exception as e:
-            app.logger.error(f"Error : {e}")
-            raise e
+        if (current_user.get_role() == UserRole.writer):
+            try:
+                status = request.args["status"]
+                app.logger.info(f"mark_as_done: id = {id}, status = {status}")
+                repository.update_task_status(id, status)
+                return redirect('/')
+            except Exception as e:
+                app.logger.error(f"Error : {e}")
+                raise e
+        else:
+            # HELP - is there a better way to handle this? how to return 401 - Unauthorized?
+            raise Exception("Unauthorized. To Change a Task Status, the User must have the 'writer' role.")
 
     @app.route('/delete/<id>', methods=['POST']) 
-    def delete(id): 
-        try:
-            app.logger.info(f"delete: id = {id}")
-            repository.delete_task(id)  
-            return redirect('/')
-        except Exception as e:
-            app.logger.error(f"Error : {e}")
-            raise e
+    @login_required
+    def delete(id):         
+        if (current_user.get_role() == UserRole.writer):
```

Comment:
We've got this same pattern in each of these three functions, if you get a chance to look at how we could turn this into a decorator it might be a nice opportunity to see how we can commonise that